Gutenpack: Break up the Jetpack preset's view bundle into separate, per-block ones.

enejb opened the pull request:
#### Changes proposed in this Pull Request

* make it so that we can build tile-galleries as a separate bundle. 


#### Testing instructions

run the usual command that builds the bundle. 
`npm run sdk gutenberg client/gutenberg/extensions/presets/jetpack -- --output-dir="../jetpack/_inc/blocks"` 

it should produce the following files.
```
editor.js
editor.css
editor.rtl.css
tiled-gallery/view.js
tiled-gallery/view.css
tiled-gallery/view.rtl.css
```

enejb committed:
Gutenpack: seperate the tiled-gallery into its own view bundle

enejb committed:
Use a new index.json preset format to generate individual blocks

enejb committed:
Remove the need for view.js and edtor.js

We are not combining all the files based on the index.json file.

simison commented on client/gutenberg/extensions/presets/jetpack/editor.js:
```diff
@@ -1,11 +0,0 @@
-import 'gutenberg/extensions/markdown/editor';
-//import 'gutenberg/extensions/publicize/editor';
-import 'gutenberg/extensions/related-posts/editor';
-import 'gutenberg/extensions/tiled-gallery/editor';
```
This preset file is imported in Calypso: 

https://github.com/Automattic/wp-calypso/blob/5c5c85f1b6cc70a6bf853ae4c7decc54a42dd75e/client/gutenberg/editor/edit-post/editor.js#L20

You might want to either keep it (and perhaps rename) or import each extension separately in `client/gutenberg/editor/edit-post/editor.js`.

enejb committed:
More cleanup

simison commented on bin/sdk/gutenberg.js:
```diff
@@ -4,20 +4,51 @@
+	  // Helps split up each block into its own folder view script 
+		viewBlocksPoints = pickBy( reduce( preset, (obj, block) => ( { [ block + '/view'] : [path.join( inputDir, '../../'+ block +'/view.js' )]  } )  ), fs.existsSync );
+
+		utilScripts = fs.readdirSync( utilsPath ).map( utilFile => path.join( utilsPath, utilFile ) ).filter( fs.existsSync );
```
I wonder if we could simply import utils in each block depending on them.

Right now building a single block is a bit tricky because many of them depend on `jetpack` category which gets set only by importing the jetpack-preset...

enejb replied:
I am just trying to make sure that we have all the same code as before.

simison replied:
Sure! Makes sense for this PR. :-)

enejb committed:
Readd the editor.js file

enejb committed:
minor fix

jsnajdr commented on bin/sdk/gutenberg.js:
```diff
@@ -4,20 +4,50 @@
+		const presetBlocks = require( presetPath );
+
+	  // Helps split up each block into its own folder view script
+		viewBlocksPoints = reduce( presetBlocks, (obj, block) => ( { [ block + '/view'] : [path.join( inputDir, '../../'+ block +'/view.js' )]  } )  );
```
What is this `reduce` supposed to do? `obj` is never used, so the whole call will return the returned value for the last block?

Maybe you want this?
```js
fromPairs( map(
  blocks,
  block => [ block + '/view', path.join( ..., block, 'view.js' ) ]
) );
```

simison commented on client/gutenberg/extensions/presets/jetpack/index.json:
```diff
@@ -0,0 +1 @@
+[ "markdown", "related-posts","tiled-gallery" ]
```
Remember to rebase, `publicize` extension [is now enabled](https://github.com/Automattic/wp-calypso/pull/28110) in Jetpack preset as well.

enejb committed:
Remove the view.js scripts since they are not needed any more.

enejb committed:
Add documention on presets.

enejb committed:
Add the utls scripts to views.js

enejb committed:
Minor change

ockham commented:
To provide a bit of rationale here (mostly discussed with @enejb and @lezama on a video call and screenshare):

The goal here is to break up the Jetpack preset's `view` bundle into separate, per-block ones, so they can be conditionally loaded by Jetpack depending on the corresponding modules' activation status -- from PHP, i.e. through `wp_enqueue_script()`.

To reach this goal, I suggested using separate webpack [entry points](https://webpack.js.org/concepts/entry-points/) for each block's `view.

We're now tasked with producing the list of those files from a given preset configuration. 

We discussed that our current preset format -- Javascript files that simply `import` individual blocks -- isn't particularly suited to this task: We would essentially need to parse JavaScript -- and even if we settled for a very basic, RegEx-based parser, those presets might also import util files that we don't want to create separate entry points for.

We also observed that our current presets configuration is somewhat redundant -- both a preset's `editor.js`, and its `view.js` essentially import each of the preset's `editor.js` and `view.js` files, respectively.

The solution that we came up with is what started this PR: Use a JSON file that only contains an array of block directory names. Iterate over this list to create both the list of view file entry points, and dynamically create an `editor.js` file in essentially the same format that we have now.

There are two tradeoffs AFAICS:

- Previously, with presets just being Javascript files that imported blocks, the SDK didn't care if it was dealing with a preset, or an individual block -- the build process would work regardless.
- We're introducing yet more conventions here, namely:
  - Block directories are expected to be at the same file system hierarchy level as the `presets/` directory. (This requirement could probably be removed by changing our preset JSONs from arrays to objects, whose keys are block names, and whose values are paths. IMO: Convention over configuration, and YAGNI.)
  - Shared util files are expected to be in the preset's `util/` subdirectory.

simison commented:
Noting that while the title mentions tiled gallery, the solution is built for all the blocks. The tiled gallery just happens to be the only one with view+editor side scripts and styles right now.

tyxla reviewed (commented):
This seems to be testing well, nice work 💯 

I've left some comments, mostly regarding code style - nothing major.

Also, I've noticed that I don't get any `view` scripts/styles in the root build directory - is that because tiled galleries is the only use case we have right now?

simison commented on bin/sdk/gutenberg.js:
```diff
@@ -4,20 +4,52 @@
+	if ( fs.existsSync( presetPath ) ) {
+		const presetBlocks = require( presetPath );
+		// Find all the utils scripts
+		utilScripts = fs.readdirSync( utilsPath ).map( utilFile => path.join( utilsPath, utilFile ) ).filter( fs.existsSync );
```
Note that in https://github.com/Automattic/wp-calypso/pull/28088 we introduced `gutenberg/extensions/presets/jetpack/utils/i18n` which is imported from each block separately, instead just once in the preset file.

Could we refactor utility-files by moving them to `/shared` and import them in each block that depends on them? That way it'll be easier to build each block separately with the SDK without relying on preset importing things such as jetpack category for us.

The purpose of the preset would thus be only to define combination of extensions; not new functionality per-se.

See also https://github.com/Automattic/wp-calypso/pull/27836 which adds `/shared` utility folder.

simison replied:
Ohh funny, looks like I commented almost identical thing in resolved convo 😆  https://github.com/Automattic/wp-calypso/pull/28066/files#r228319972 — sorry I missed that.

Introduction of `utils/i18n` changes the situation again quite a bit, though.

simison commented:
Many of the styling issues seem to be coming from the fact that `bin` is ignored by `eslint`. :-(

https://github.com/Automattic/wp-calypso/blob/ba89961513119449206fd12b5df3d931661df7ff/.eslintignore#L3

You might want to run `npx eslint bin/sdk/gutenberg.js  --no-ignore` manually to make sure all is good.

simison commented:
By running:
```
npm run sdk -- gutenberg client/gutenberg/extensions/presets/jetpack/ --output-dir=~/jetpack/_inc/blocks/
```

I'm getting:
```
_inc/blocks/
├── editor.css
├── editor.js
├── editor.rtl.css
└── tiled-gallery
    ├── view.css
    ├── view.js
    └── view.rtl.css
```

The PR description doesn't mention https://github.com/Automattic/jetpack/pull/10405 but I'm assuming these both should be merged at the same time?

Heads up @oskosk that Jurassic Ninja might need `master` branch of Jetpack for each build for a while again, because asset paths have changed.

simison reviewed (dismissed):
Building the `o2` preset seems to be broken right now:

```bash
npm run sdk -- gutenberg client/gutenberg/extensions/presets/o2 --output-dir=~/test
```

As well as building individual blocks:
```bash
npm run sdk -- gutenberg client/gutenberg/extensions/tiled-gallery --output-dir=~/test
```

Building via `bin` script seems to work well — to test, type `npm link` in Calypso folder and then from Jetpack folder run:

```
calypso-sdk gutenberg ~/wp-calypso/client/gutenberg/extensions/presets/jetpack --output-dir=./_inc/blocks
```

enejb committed:
Fixes suggested by Marin and Mikael

enejb committed:
Update and add README.md

enejb commented on client/gutenberg/extensions/presets/jetpack/index.json:
```diff
@@ -0,0 +1 @@
+[ "markdown", "related-posts","tiled-gallery", "publicize" ]
```
:+1:

enejb commented on bin/sdk/gutenberg.js:
```diff
@@ -4,20 +4,52 @@
+		editorScripts = presetBlocks.map( block => path.join( inputDir, '../../'+ block +'/editor.js' ) ).filter( fs.existsSync );
+		viewScripts = presetBlocks.map( block => path.join( inputDir, '../../'+ block +'/view.js' ) ).filter( fs.existsSync );
+		// Combines all the different blocks into one Edit script
+		editorScript = utilScripts.concat( editorScripts ).concat( viewScripts );
```
:+1:

enejb commented on bin/sdk/gutenberg.js:
```diff
@@ -4,20 +4,52 @@
+		forEach( presetBlocks, block => {
+				const viewScriptPath = path.join( inputDir, '../../'+ block +'/view.js' );
+				if ( fs.existsSync( viewScriptPath ) ) {
+					viewBlocksPoints = Object.assign( viewBlocksPoints, { [ block + '/view' ] : utilScripts.concat( [  viewScriptPath ] ) } );
```
:+1:

enejb committed:
Use string litteral

enejb commented on bin/sdk/gutenberg.js:
```diff
@@ -4,20 +4,52 @@
+
+		viewBlocksPoints = {}
+		// Helps split up each block into its own folder view script
+		forEach( presetBlocks, block => {
```
:+1:

enejb commented:
@tyxla and @simison 
I updated the PR so that we don't create view.js and view.css files. 

```
editor.js
editor.css
editor.rtl.css
tiled-gallery/view.js
tiled-gallery/view.css
tiled-gallery/view.rtl.css
```

Since view.css and js files should just be included in the editor. (js and css) files. No need to load them separately.

I also created a `shared` that contains the js files that should be included with each bundle. 
and editor-shared folder that should contain only the editor shared includes.

enejb committed:
Fix editor.js includes

simison committed:
Run Prettier on files

```
npx prettier --write \
  bin/sdk/gutenberg.js \
  docs/sdk.md \
  client/gutenberg/extensions/presets/jetpack/utils/README.md
```

simison commented on bin/sdk/gutenberg.js:
```diff
@@ -2,6 +2,8 @@
  * External dependencies
  */
+
+/* eslint import/no-nodejs-modules: ["error", {"allow": ["path", "fs"]}] */
```
Noting that I made this rule global for `/bin` files in https://github.com/Automattic/wp-calypso/pull/28167 but no changes needed for now here.

simison reviewed (approved):
I tested different ways to use the SDK (building all different presets, building a single block, building using the npm script and the bin command) and everything worked as expected.

Nice job driving this pretty big change through while keeping backwards compatibility! :thumbsup:

tyxla commented on bin/sdk/gutenberg.js:
```diff
@@ -10,14 +12,65 @@ exports.config = ( { argv: { inputDir, outputDir }, getBaseConfig } ) => {
+	let viewBlocksScripts;
+	let viewScriptEntry;
+	if ( fs.existsSync( presetPath ) ) {
+		const DIRECETORY_DEPTH = '../../';
```
Typo here: `DIRECETORY_DEPTH`

tyxla commented on bin/sdk/gutenberg.js:
```diff
@@ -10,14 +12,65 @@ exports.config = ( { argv: { inputDir, outputDir }, getBaseConfig } ) => {
+		const presetBlocks = require( presetPath );
+		// Find all the shared scripts
+		const sharedPath = path.join( inputDir, 'shared' );
+		const sharedUtilsScripts = fs
```
We seem to duplicate code for similar operations here - we could definitely pluck out some of it in helper functions.

tyxla reviewed (approved):
Test well for me 👍 Left a couple of minor suggestions, other than that LGTM 🚢

enejb committed:
Extact some functsion to make the code more DIY
